### DISCUSSION TITLE
Errors when compiling st-flexipatch

### DISCUSSION THREAD

**Discussion by bakkeby** [bakkeby/st-flexipatch#25] [Known Issues]
Some patches depend on additional libraries which:
   - need to be installed and
   - corresponding lines to include said libraries will need to be uncommented in the [config.mk](https://github.com/bakkeby/st-flexipatch/blob/master/config.mk) file

There are usually a note about what libraries are required in the [patches.h](https://github.com/bakkeby/st-flexipatch/blob/master/patches.def.h).

### Example errors

```make
/usr/bin/ld: x.o: in function `xinit':
x.c:(.text+0x337d): undefined reference to `XcursorLibraryLoadCursor'
collect2: error: ld returned 1 exit status
make: *** [Makefile:33: st] Error 1
```

Caused by enabling the [themed cursor](https://st.suckless.org/patches/themed_cursor/) patch (THEMED_CURSOR_PATCH) and forgetting to uncomment this:
https://github.com/bakkeby/st-flexipatch/blob/5adf4c4c8eaad4535fa99cbd905daec4b22f6939/config.mk#L18-L19

Requires the libxcursor / Xcursor library.

---

```make
/usr/bin/ld: x.o: in function `xmakeglyphfontspecs':
x.c:(.text+0x98b): undefined reference to `hbtransform'
/usr/bin/ld: x.o: in function `zoomabs':
x.c:(.text+0x2d88): undefined reference to `hbunloadfonts'
collect2: error: ld returned 1 exit status
make: *** [Makefile:33: st] Error 1
```

Caused by enabling the [ligatures](https://st.suckless.org/patches/ligatures/) patch (LIGATURES_PATCH) and forgetting to uncomment this:
https://github.com/bakkeby/st-flexipatch/blob/5adf4c4c8eaad4535fa99cbd905daec4b22f6939/config.mk#L21-L25

Requires the harfbuzz library.

---

```make
/usr/bin/ld: st.o: in function `strhandle':
st.c:(.text+0x545c): undefined reference to `sixel_parser_deinit'
/usr/bin/ld: st.c:(.text+0x577e): undefined reference to `sixel_parser_finalize'
/usr/bin/ld: st.c:(.text+0x5794): undefined reference to `sixel_parser_deinit'
/usr/bin/ld: st.o: in function `tputc':
st.c:(.text+0x5930): undefined reference to `sixel_parser_parse'
/usr/bin/ld: st.c:(.text+0x6428): undefined reference to `sixel_parser_init'
collect2: error: ld returned 1 exit status
make: *** [Makefile:33: st] Error 1
```

Caused by enabling the [sixel](https://gist.github.com/saitoha/70e0fdf22e3e8f63ce937c7f7da71809) patch (SIXEL_PATCH) and forgetting to uncomment this:
https://github.com/bakkeby/st-flexipatch/blob/5adf4c4c8eaad4535fa99cbd905daec4b22f6939/config.mk#L27-L28

Requires no additional libraries.

**Comment by kmarzic** [comment]
Hi,

WIth included patch `#define NETWMICON_PATCH 1` I am getting the following compile problem:

```
$ diff patches.def.h patches.h  | grep "^>"
> #define NETWMICON_PATCH 1
```

```
$ make
c99 -I/usr/X11R6/include  `pkg-config --cflags fontconfig`  `pkg-config --cflags freetype2`   -DVERSION=\"0.9\" -DICON=\"/usr/local/share/pixmaps/st.ff\" -D_XOPEN_SOURCE=600  -march=core2 -mtune=generic -pipe -c st.c
c99 -I/usr/X11R6/include  `pkg-config --cflags fontconfig`  `pkg-config --cflags freetype2`   -DVERSION=\"0.9\" -DICON=\"/usr/local/share/pixmaps/st.ff\" -D_XOPEN_SOURCE=600  -march=core2 -mtune=generic -pipe -c x.c
c99 -o st st.o x.o -L/usr/X11R6/lib -lm -lrt -lX11 -lutil -lXft    `pkg-config --libs fontconfig`  `pkg-config --libs freetype2`
/bin/ld: x.o: in function `setnetwmicon':
x.c:(.text+0x5a): undefined reference to `gdImageCreateFromPng'
/bin/ld: x.c:(.text+0x1a7): undefined reference to `gdImageGetPixel'
/bin/ld: x.c:(.text+0x2ed): undefined reference to `gdImageDestroy'
collect2: error: ld returned 1 exit status
make: *** [Makefile:27: st] Error 1
```

I am not 100% sure where is the problem (somewhere on my side, or in the code).

Best regards

**Reply by step-** [reply]
The NETWMINCON_PATCH was split into three separate patches recently. As a result, NETWMICON_PATCH got a new requirement (libgd) in config.mk. You need to uncomment some code in config.mk. Note that if your intent is to compile st-flexipatch with the built-in icon, and not  with an external PNG icon, you need to enable NETWMICON_LEGACY_PATCH instead of NETWMICON_PATCH.

**Reply by kmarzic** [reply]
Thanks!
It is working, my mistake :(
